projects/driver_controls/src/pedal_calibration.c

```diff
@@ -0,0 +1,195 @@
+      (min_brake + max_brake) / 2;
+  throttle_calibration->line[THROTTLE_CHANNEL_MAIN].full_throttle_reading =
+      (min_accel + max_accel) / 2;
+
+  // Get the full range for main channel. Range is useful since when multiplied by zone percentages,
+  // it will give the size of each zone.
+  int16_t range = storage->band[main_channel][PEDAL_CALIBRATION_STATE_FULL_THROTTLE].max -
+                  storage->band[main_channel][PEDAL_CALIBRATION_STATE_FULL_BRAKE].min;
+
+  // Use the percentages to calculate the thresholds of each zone.
+  ThrottleZoneThreshold *zone_thresholds = throttle_calibration->zone_thresholds_main;
+
+  zone_thresholds[THROTTLE_ZONE_BRAKE].min = min_brake;
+  zone_thresholds[THROTTLE_ZONE_BRAKE].max =
+      zone_thresholds[THROTTLE_ZONE_BRAKE].min + (brake_zone_percentage * range / 100);
+
+  zone_thresholds[THROTTLE_ZONE_COAST].min = zone_thresholds[THROTTLE_ZONE_BRAKE].max + 1;
+  zone_thresholds[THROTTLE_ZONE_COAST].max =
+      zone_thresholds[THROTTLE_ZONE_COAST].min + (coast_zone_percentage * range / 100);
+
+  zone_thresholds[THROTTLE_ZONE_ACCEL].min = zone_thresholds[THROTTLE_ZONE_COAST].max + 1;
+  zone_thresholds[THROTTLE_ZONE_ACCEL].max = max_accel;
+
+  // Get the secondary band's vertices.
+  min_brake = storage->band[secondary_channel][PEDAL_CALIBRATION_STATE_FULL_BRAKE].min;
+
+  max_brake = storage->band[secondary_channel][PEDAL_CALIBRATION_STATE_FULL_BRAKE].max;
+
+  min_accel = storage->band[secondary_channel][PEDAL_CALIBRATION_STATE_FULL_THROTTLE].min;
+
+  max_accel = storage->band[secondary_channel][PEDAL_CALIBRATION_STATE_FULL_THROTTLE].max;
+
+  // Calculate the secondary line as a "midline" of the secondary band.
+  throttle_calibration->line[THROTTLE_CHANNEL_SECONDARY].full_brake_reading =
+      (min_brake + max_brake) / 2;
+  throttle_calibration->line[THROTTLE_CHANNEL_SECONDARY].full_throttle_reading =
+      (min_accel + max_accel) / 2;
+  // Tolerance should be half of the band's width assuming the width is constant.
+  // In this case we take the maximum of widths at both ends multiplied by the given safety factor.
+  throttle_calibration->tolerance =
+      MAX((max_brake - min_brake) / 2, (max_accel - min_accel) / 2) * tolerance_safety_factor;
+
+  return STATUS_CODE_OK;
+}
+
+// Returns main or secondary ADS1015 channels.
+// This prevents mixing the channels when calling throttle_init.
+// Should be called after calling pedal_calibration_calculate.
+Ads1015Channel pedal_calibration_get_channel(PedalCalibrationStorage *storage,
+                                             ThrottleChannel channel) {
+  if (channel == THROTTLE_CHANNEL_MAIN) {
+    return storage->adc_channel[PEDAL_CALIBRATION_CHANNEL_A];
+  } else if (channel == THROTTLE_CHANNEL_SECONDARY) {
+    return storage->adc_channel[PEDAL_CALIBRATION_CHANNEL_B];
+  } else {
+    return NUM_ADS1015_CHANNELS;
+  }
+}
+
+StatusCode pedal_calibration_validate(PedalCalibrationStorage *storage,
```

Comment:
This will just block forever. You have no way of exiting. Why not just integrate this into your test/FSM?